stellargraph/core/element_data.py

```diff
@@ -331,28 +364,53 @@ def __init__(self, shared, type_starts):
 
     def _init_directed_adj_lists(self):
         # record the edge ilocs of incoming, outgoing and both-direction edges
-        in_dict = {}
-        out_dict = {}
+        if len(self.targets) == 0:
+            empty_array = np.array([], dtype=np.uint8)
+            self._edges_in_dict = FlatAdjacencyList(empty_array, empty_array)
+            self._edges_out_dict = FlatAdjacencyList(empty_array, empty_array)
+            return
+
+        # this could be less than the true number of nodes if the node with the largest iloc is isolated
+        # but this doesn't cause issues with the code below
+        number_of_nodes = max(self.targets.max(), self.sources.max()) + 1
+        dtype = np.min_scalar_type(len(self.sources))
 
-        for i, (src, tgt) in enumerate(zip(self.sources, self.targets)):
-            in_dict.setdefault(tgt, []).append(i)
-            out_dict.setdefault(src, []).append(i)
+        flat_array = np.argsort(self.targets)
+        node_ilocs, counts = rle(self.targets[flat_array])
+        in_neighbour_counts = np.zeros(number_of_nodes, dtype=dtype)
+        in_neighbour_counts[node_ilocs] = counts
+        self._edges_in_dict = FlatAdjacencyList(flat_array, in_neighbour_counts)
 
-        dtype = np.min_scalar_type(len(self.sources))
-        self._edges_in_dict = _numpyise(in_dict, dtype=dtype)
-        self._edges_out_dict = _numpyise(out_dict, dtype=dtype)
+        flat_array = np.argsort(self.sources)
+        node_ilocs, counts = rle(self.sources[flat_array])
+        out_neighbour_counts = np.zeros(number_of_nodes, dtype=dtype)
+        out_neighbour_counts[node_ilocs] = counts
+        self._edges_out_dict = FlatAdjacencyList(flat_array, out_neighbour_counts)
 
     def _init_undirected_adj_lists(self):
         # record the edge ilocs of incoming, outgoing and both-direction edges
-        undirected = {}
-
-        for i, (src, tgt) in enumerate(zip(self.sources, self.targets)):
-            undirected.setdefault(tgt, []).append(i)
-            if src != tgt:
-                undirected.setdefault(src, []).append(i)
-
-        dtype = np.min_scalar_type(len(self.sources))
-        self._edges_dict = _numpyise(undirected, dtype=dtype)
+        num_edges = len(self.targets)
+        if num_edges == 0:
+            empty_array = np.array([], dtype=np.uint8)
+            self._edges_dict = FlatAdjacencyList(empty_array, empty_array)
+            return
+
+        number_of_nodes = max(self.targets.max(), self.sources.max()) + 1
+        dtype = np.min_scalar_type(num_edges)
+        sentinel = np.cast[self.sources.dtype](-1)
+
+        combined = np.concatenate([self.sources, self.targets])
+        # mask out duplicates of self loops
+        # can't remove because this would invalidate the argsort results
+        combined[num_edges:][self.sources == self.targets] = sentinel
```

Comment:
Rather than use a sentinel, maybe we could do a little more work to filter `self.targets` and not rely on the `% num_edges` later.

```python
non_self_loops = self.source != self.targets
edge_ilocs = np.concatenate([np.arange(num_edges), np.where(non_self_loops)[0]])
combined = np.concatenate([self.sources, self.targets[non_self_loops])

sorting = np.argsort(combined)
flat_array = edge_ilocs[sorting]
node_ilocs, counts = rle(combined[sorting])

...
```

I feel like this is vaguely clearer?

Reply:
unfortunately this approach breaks because we change the the args of `self.targets[non_self_loops]`  so they no long correspond to edge ilocs

e.g.:

```
sources = [0, 1, 2]
targets = [0, 2, 1]
combined = [0, 1, 2, 2, 1]
sorting = [0, 1, 2, 3, 4]
sorting % num_edges = [0, 1, 2, 0, 1]
```

So this incorrectly associates the edge ilocs of 0 and 1 instead othe correct 1 and 2 - if that makes sense??

Reply:
I think the `np.where(non_self_loops)[0]` handles that. A (more expensive) alternative that more directly expresses the behaviour would be `np.arange(num_edges)[non_self_loop]`.

Reply:
oh my bad I read your initial code wrong!

this uh might be hypocritical because i feel like there's some extra arrays i could cut down on, but this creates an extra array of size `E`. Although it does make the code alot clearer I'm inclined to say its not worth the extra construction memory cost. What do you think?

Reply:
> this uh might be hypocritical because i feel like there's some extra arrays i could cut down on, but this creates an extra array of size E. 

Yeah, like `flat_array % num_edges` is creating a whole new array of size E. And in a super-expensive way, since `%` is an incredibly slow arithmetic operation:

```python
x = np.arange(10000)

%timeit -n 10000 x % 10
# 82.4 µs ± 1.45 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit -n 10000 x + 10
# 2.86 µs ± 90.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Reply:
> flat_array % num_edges is creating a whole new array of size E.

haha yeah good point! I've changed this to be inplace. Re the extra time for `%` for larger edges lists I reckon this will be neglible compared to the cost of sorting.

Using `bincount` + extra sorting I think i've cut down the extra arrays to just the `combined` array and a binary `self_loops` array. The extra sort likely doubles the undirected constructor time at the expense of saving memory. After reading the convo in the data science team I feel like memory saving might be more important. Thoughts?

Also I reckon in the future we could avoid even constructing the `combined`  array by storing `sources` and `targets` in one contiguous array + some extra tricks

Reply:
> Re the extra time for % for larger edges lists I reckon this will be neglible compared to the cost of sorting.

It's not negligible for moderate arrays, but it is at least O(n) whereas the sorting will of course be O(n log n), and so, it is much faster for larger arrays.

```python
import numpy as np
N = 100000
arr = np.random.permutation(N)
%timeit -n 100 np.sort(arr)
%timeit -n 100 arr % 54321
```

| N | sort | `%` |
|---|---|---|
| 1000 | 27us ± 3.6 | 12.3us ± 0.954 |
| 100000 | 4750us ± 165 | 770us ± 18.4 |
| 10000000 | 686ms ± 20 | 80.7ms ± 3.79 |

(I reduced it to `-n 1` for the last row).